src/avro-derive/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright Materialize, Inc., and other contributors (if applicable)
+//
+// Use of this software is governed by the Apache License, Version 2.0
+
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+use syn::parse_macro_input;
+use syn::ItemStruct;
+
+#[proc_macro_derive(AvroDecodeable, attributes(decoder_factory))]
+pub fn derive_decodeable(item: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(item as ItemStruct);
+    let name = input.ident;
+    let fields: Vec<_> = input
+        .fields
+        .iter()
+        .map(|f| {
+            let ty = &f.ty;
+            let id = f.ident.as_ref().unwrap();
+            quote! {
+                #id: Option<#ty>
+            }
+        })
+        .collect();
+
+    let decode_blocks: Vec<_> = input
+        .fields
+        .iter()
+        .map(|f| {
+            let ty = &f.ty;
+            let id = f.ident.as_ref().unwrap();
+            let id_str = id.to_string();
+            let found_twice = format!("field `{}` found twice", id);
+            let make_decoder =
+                if let Some(decoder_factory) = f.attrs.iter().find(|a| {
+                    &a.path.get_ident().as_ref().unwrap().to_string() == "decoder_factory"
+                }) {
+                    let toks = &decoder_factory.tokens;
+                    quote! {
+                        #toks()
+                    }
+                } else {
+                    quote! {
+                        <#ty as ::mz_avro::AvroDecodeable>::new_decoder()
+                    }
+                };
+            quote! {
+                #id_str => {
+                    if self.#id.is_some() {
+                        ::anyhow::bail!(#found_twice);
+                    }
+                    let decoder = #make_decoder;
+                    self.#id = Some(field.decode_field(decoder)?);
+                }
+            }
+        })
+        .collect();
+    let check_blocks: Vec<_> = input
+        .fields
+        .iter()
+        .map(|f| {
+            let id = f.ident.as_ref().unwrap();
+            let not_found = format!("field `{}` not found", id);
+            quote! {
+                let #id = if let Some(#id) = self.#id.take() {
+                    #id
+                } else {
+                    ::anyhow::bail!(#not_found);
+                };
+            }
+        })
+        .collect();
+    let return_fields: Vec<_> = input
+        .fields
+        .iter()
+        .map(|f| f.ident.as_ref().unwrap())
+        .collect();
+    let decoder_name = format_ident!("{}_DECODER", name);
+    let out = quote! {
+        #[derive(Default)]
+        #[allow(non_camel_case_types)]
+        struct #decoder_name {
+            #(#fields),*
+        }
+        impl ::mz_avro::AvroDecode for #decoder_name {
```

Comment:
You should document that this proc macro requires mz_avro be linked with this name. I believe that there is a way to make this more robust by re-exporting from this crate, but I don't recall the details and it can be dealt with in the future if it's a problem.